### ISSUE TITLE
Trying to make request from App Extension

### ISSUE THREAD

**Issue by eytanschulman** [google/google-api-objectivec-client#159]
I'm trying to make a Google Calendar request from an app extension, and it does not compile as GTMHTTPFetcher references UIApplication (which is not accessible from an app extension).

Is there a way to use the library in an app extension?

**Comment by thomasvl** [comment]
Current versions of the library (with current SDKs/Xcodes) should default to using NSURLSession.  I think that code path is safe for Extensions.

**Comment by eytanschulman** [comment]
@thomasvl I'm using CocoaPods and because of that, gam-http-fetcher and gtm-oauth2 get added in addition to the Google-API-Client. Because of that I get an error on this line accessing UIApplication:
`UIApplication *app = [UIApplication sharedApplication];` saying "GTMHTTPFetcher.m:483:41: 'sharedApplication' is unavailable: not available on iOS (App Extension) - Use view controller based solutions where appropriate instead."

**Comment by thomasvl** [comment]
I'm not sure what what pod you are using, there isn't an official one; we're finishing up landing one: https://github.com/google/google-api-objectivec-client/pull/157

**Comment by eytanschulman** [comment]
@thomasvl Thanks, didn't know that [this](https://cocoapods.org/pods/Google-API-Client) wasn't official but now I see that the authors doesn't say "Google". Does CocoaPods have to manually add the new pod for it to get up?

**Comment by thomasvl** [comment]
I'm relatively clueless on a lot of the pod process.

@perotinus - Do we need to do anything to start getting the pods landed pickup by the general cocoapod ecosystem?

**Comment by eytanschulman** [comment]
@thomasvl After some [research](https://guides.cocoapods.org/making/getting-setup-with-trunk.html), it seems as if the creator of the pod (on your end), will need to upload it to CocoaPods's Trunk API for it to be publicly available.

**Comment by perotinus** [comment]
I believe this has been done. See #157 for more details.

**Comment by eytanschulman** [comment]
@perotinus Thanks, it has been done and i've removed integration with the old, unofficial pod and added the new one. It does not fix the issue with using it in an extension as the UIApplication calls are now in GTMSessionFetcher as opposed to GTMHTTPFetcher.

**Comment by perotinus** [comment]
@eytanschulman It looks like the functionality that uses UIApplication is gated by a macro, `GTM_BACKGROUND_TASK_FETCHING`. If you add a `post_install` hook that sets this to `0`, does that fix your issue?

Something like:
```
post_install do |installer|
  installer.pods_project.targets.each do |target|
    target.build_configurations.each do |config|
      config.build_settings['GTM_BACKGROUND_TASK_FETCHING'] = 0
    end
  end
end
```
but more tuned to only apply to GTMSessionFetcher?

**Comment by eytanschulman** [comment]
@perotinus I've tried to both add that snippet to my podfile and manually 'GTM_BACKGROUND_TASK_FETCHING=0' in my extension's Preprocessor Macros to no avail.

**Comment by perotinus** [comment]
Ah, OK. My mistake: what I gave you above creates a build setting, but not a preprocessor definition. Try this:
```
post_install do |installer|
  installer.pods_project.targets.each do |target|
    target.build_configurations.each do |config|
      config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] = '$(inherited) GTM_BACKGROUND_TASK_FETCHING=0'
    end
  end
end
```
I tried this out locally, and that macro was defined when compiling, though I didn't test in an extension.

Changing the settings in your extension wouldn't work, because those settings only apply to code compiled in your extension's target. You would have to go to the Pods project, the GTMSessionFetcher target and make the change for it to affect the GTMSessionFetcher code.

**Comment by eytanschulman** [comment]
@perotinus Thanks! It worked. Now there are other errors that were hiding until those previous ones were solved.. For some reason the new pod is giving me "Unknown type name 'GTMHTTPFetcherService'" on this line: `GTMBridgeFetcherService *fetcherService_; `in GTLService.h

**Comment by perotinus** [comment]
@eytanschulman Which SDK version are you compiling against? If you're using the iOS 9 SDK, this should default to GTMSessionFetcher which I think will cause another problem, which we're working on :#162). If not, you will have to define that flag by hand in the post_install or by modifying the project.

**Comment by eytanschulman** [comment]
@perotinus Base SDK: 9.2, Deployment Target: 8

**Comment by perotinus** [comment]
@eytanschulman The GTMOAuth2 podspec is now live. Can you try that out and see if it fixes the issues with GTMOAuth2?

I'm still not sure why you're getting the errors about GTMHTTPFetcher. If you define GTM_USE_SESSION_FETCHER to 1 in the preprocessor flags, or at the top of GTLService.h, do the GTMHTTPFetcher errors go away? Did you define GTM_USE_SESSION_FETCHER somewhere else?

**Comment by eytanschulman** [comment]
@perotinus Thanks, this solves all **compile time** errors I was getting!
Now i'm just trying to work out GTMOAuth2 on my NCWidget. I opened google/gtm-oauth2/#59

**Comment by jonblatho** [comment]
Updating this thread because the solution given by @perotinus has changed with a [new pull request](https://github.com/google/gtm-session-fetcher/pull/70) with a different preprocessor flag.

Just take the lines given by his comment and replace `GTM_BACKGROUND_TASK_FETCHING` with `GTM_BACKGROUND_UIAPPLICATION`...and it should clear things up, if you tried this solution recently and ran into issues like I did.
